src/virtual_machine.rs

```diff
@@ -34,6 +77,113 @@ impl VM {
             rcount: 0,
         }
     }
+
+    fn decode_instruction(instruction: u16) -> Opcode {
+        // Removes the arguments from the instruction, leaving only the operator
+        let op = instruction >> ARGSIZE;
+        // Removes the operator from the instruction, leaving only the arguments
+        let args = instruction & ARGUMENTMASK;
+
+        // NOTE: Whilst writing constants inside a code block
+        // is not very orthodox, I believe it remains a good
+        // balance between avoiding magic numbers and keeping
+        // the logic "local".
+        match op {
+            // BR
+            0b0000 => {
+                todo!()
+            }
+            // ADD
+            0b0001 => {
+                const ADDMODEPOSITION: u16 = 5;
+                const ADDMODEFLAG: u16 = 0b0000000000100000;
+                let mode = (args & ADDMODEFLAG) >> ADDMODEPOSITION;
+
+                test_print(format!("{:?}", mode).as_str());
+                let second_arg = match mode {
+                    0 => {
+                        const IMMEDIATEREGISTERPOS: u16 = 0b0000000000000111;
+                        let dest_register = args & IMMEDIATEREGISTERPOS;
+                        AddMode::IMMEDIATE { sr2: dest_register }
+                    }
+                    1 => {
+                        todo!()
+                    }
+                    _ => panic!("ERROR WHILST PARSING"),
+                };
+
+                const DESTINATIONREG: u16 = 0b0000111000000000;
+                const DESTINATIONREGPOS: u16 = 9;
+                let dr = (args & DESTINATIONREG) >> DESTINATIONREGPOS;
+
+                const SOURCEREG: u16 = 0b0000000111000000;
+                const SOURCEREGPOS: u16 = 6;
+                let sr1 = (args & SOURCEREG) >> SOURCEREGPOS;
```

Comment:
Same comment, this adds overhead, I would just write directly the number if you are only going to use it once.